java/src/main/native/src/RmmJni.cpp

```diff
@@ -26,12 +26,120 @@ using rmm::mr::device_memory_resource;
 
 namespace {
 
+// Alignment to which the RMM memory resource will round allocation sizes
+constexpr std::size_t RMM_ALLOC_SIZE_ALIGNMENT = 512;
+
 constexpr char const* RMM_EXCEPTION_CLASS = "ai/rapids/cudf/RmmException";
 
+/**
+ * @brief An RMM device memory resource that delegates to another resource
+ * while tracking the amount of memory allocated.
+ *
+ * @tparam Upstream Type of memory resource that will be wrapped.
+ * @tparam size_align The size to which all allocation requests are
+ * aligned. Must be a value >= 1.
+ */
+template <typename Upstream>
```

Comment:
`tracking_resource_adaptor` is not a public class for public consumption.  It is private to this compilation unit, and in the future we have no intention of trying to use it with private memory resources.  SFINAE here seems like it's a solution looking for a problem to solve.

Reply:
Honestly there is not even much point in using a template here except possibly for flexibility in the future because we instantiate this once and only once with `Upstream` set to `device_memory_resource`.  In other similar places the template is used to try and eliminate vtable lookups, and we may do the same once we combine this with #4816.

Reply:
Yes, the only reasons I used a template argument here was because I could see the #4816 changes coming in soon, and with that we would know which specific resource type(s) would be used underneath.  It doesn't make sense with this change alone. I just put it there in anticipation of the resource setup code coming in #4816.

Reply:
> It is private to this compilation unit...
Ah, yes, I missed that. Thanks.